### PR TITLE
Orderers heavy cleanup

### PR DESCRIPTION
**Pull request by ranchalp**
This PR is a quite heavy clean up (refactorization) of the orderers module. 
In short: 

- `internal/parts` holds the protocol logic, separated in three subparts (`viewchange, goodcase and catchup`)
- `internal/common` holds state that is internally common to all the three subparts of the module 
- `common` holds state that is common to both the module and used outside the module

Note that this partitioning can be subject to modifications, but there are limitations once one starts subdividing module logic (to avoid cyclic imports). 

Anticipating a possible comment: I initially tried to inline all functions for easier readability (similarly to the availability module) by removing redundant calls to auxiliary functions in the `Upon...` methods. Once I completed this, I am left with the doubt of whether this is always more readable than having `apply...` auxiliary functions as only body of the `Upon...` methods. I think that if the `Upon...` is easy to follow and part of the core logic then leaving it inline makes sense, but otherwise can be left to be treated with an auxiliary function.

In fact, inlining all of this has left the viewchange logic so big that there is a lint error (reason for which lint fails currently). So this is very much open for the reviewer to discuss (as many other things surely).

P. S. : an artifact of the squash is 3 small commits that do very little (nothing?) to the code. There is really only one massive commit in this PR (my apologies, but the refactorization was too big a change to make meaningful commits without leaving weird blocks on both ends of the process, othen than I should have inlined functions in a different commit)

### CODE REVIEW

**Review by matejpavlovic — CHANGES_REQUESTED** [review]
Thanks for the heavy lifting here!
All directly requested changes are of cosmetic character.
I left some other comments, on the other hand, that request changes best addressed in separate PRs. So feel free to just create corresponding issues and not modify the code for now.

Regarding the complexity of the function handlers that are, technically, all in a single go function, I see your point and I agree that currently it is sub-optimal. Here's my suggestion:

Let's pick one way of approaching this per file. Either
1. everything is in-line, i.e., inside the `Upon...` handler registration call, as an anonymous function, the way simple stuff like the the availability module (auxiliary functions are still encouraged, but the high-level logic of a handler would be directly inside the handler), or
2. all handler registrations would be one-liners, simply referring to auxiliary handler functions *of the same name*$^1$ (but with a lower-case first letter, to not export them). This to manually writing the code that I thought eventually could be generated (stubs of those auxiliary functions). What do you think?

$^1$ Unless 2 Handlers of the same name are defined in different packages and both used.

File: pkg/trantor/system.go [hunk 1]
```diff
@@ -147,6 +148,7 @@ func New(
 	}
 	trantorModules[moduleConfig.ISS] = issProtocol
 
+	//<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/trantor/system.go:151]: Oh yeah I just commented part of the rebase there cause I was not sure at the time which o the two would fit. I bit that comment though. The rebasing went well, not too complicated anyways.

File: pkg/orderers/internal/parts/goodcase/pbftgoodcase.go
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/parts/goodcase/pbftgoodcase.go:1]: We used to but you are right that it is not the case for `Propose` anymore. In any case, recall that exported functions in subfolders of `'module'/internal` can only be used by packages in the folder `'module'` and their submodules.

I will now go through each of the functions and verify whether they need to be exported or not here.

File: pkg/orderers/internal/parts/goodcase/pbftgoodcase.go [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+// ============================================================
+// Data structures and functions
+// ============================================================
+
+// ============================================================
+// OrdererEvent handling
+// ============================================================
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/parts/goodcase/pbftgoodcase.go:42]: Done

File: pkg/orderers/internal/common/pbftviewchangestate.go [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+	// (in the current implementation, the fresh value is a pre-configured proposal
+	// or, if none has been specified, an empty byte slice).
+	Reproposals map[tt.SeqNr][]byte
+
+	// Preprepare messages to be reproposed in the new view.
+	// At initialization, an explicit nil entry is created for each sequence number of the segment.
+	// Based on received ViewChange messages, each value will eventually be set to
+	// a new Preprepare message to be re-proposed (with a correctly set view number).
+	// This also holds for sequence numbers for which nothing was prepared in the previous view,
+	// in which case the value is set to a Preprepare with an empty certificate and the "aborted" flag set.
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/common/pbftviewchangestate.go]: done

File: pkg/orderers/internal/common/common.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+	// It is set on sending the ViewChange message and cleared on accepting a new view.
+	InViewChange bool
+
+	// For each view transition, stores the state of the corresponding PBFT view change sub-protocol.
+	// The map itself is allocated on creation of the pbftInstance, but the entries are initialized lazily,
+	// only when needed (when the node initiates a view change).
+	ViewChangeStates map[ot.ViewNr]*PbftViewChangeState
+}
+
+// pbftProposalState tracks the state of the pbftInstance related to proposing certificates.
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/common/common.go:68]: done.

File: pkg/orderers/internal/common/common.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+	// PBFT-specific configuration parameters (e.g. view change timeout, etc.)
+	Config *PBFTConfig
+
+	// Contains the application-specific code for validating incoming proposals.
+	ExternalValidator ValidityChecker
+}
+
+type State struct {
+
+	// The segment governing this SB instance, specifying the leader, the set of sequence numbers, the buckets, etc.
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/common/common.go:35]: Removed

File: pkg/orderers/internal/parts/goodcase/pbftgoodcase.go [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+// and does not perform any checks in this regard.
+func Propose(m dsl.Module, state *common.State, params *common.ModuleParams, moduleConfig common2.ModuleConfig, data []byte, logger logging.Logger) error {
+
+	// Update proposal counter.
+	sn := state.Segment.SeqNrs()[state.Proposal.ProposalsMade]
+	state.Proposal.ProposalsMade++
+
+	// Log debug message.
+	logger.Log(logging.LevelDebug, "Proposing.",
+		"sn", sn)
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/parts/goodcase/pbftgoodcase.go:233]: Done

File: pkg/orderers/internal/parts/goodcase/pbftgoodcase.go [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+		}
+		applyMsgCommit(m, state, params, moduleConfig, commit, from, logger)
+		return nil
+	})
+
+}
+
+// applyProposeTimeout applies the event of the proposal timeout firing.
+// It updates the proposal state accordingly and triggers a new proposal if possible.
+func applyProposeTimeout(m dsl.Module, state *common.State, params *common.ModuleParams, moduleConfig common2.ModuleConfig, numProposals int, logger logging.Logger) error {
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/parts/goodcase/pbftgoodcase.go:179]: Agreed, `common2` was an artifact of automatic refactoring. Had in mind to fix that in a future PR. 

Fixing the headers now in this PR

**Comment by ranchalp** [comment]
> Let's pick one way of approaching this per file. Either
> 
>     1. everything is in-line, i.e., inside the `Upon...` handler registration call, as an anonymous function, the way simple stuff like the the availability module (auxiliary functions are still encouraged, but the high-level logic of a handler would be directly inside the handler), or
> 
>     2. all handler registrations would be one-liners, simply referring to auxiliary handler functions _of the same name_1 (but with a lower-case first letter, to not export them). This to manually writing the code that I thought eventually could be generated (stubs of those auxiliary functions). What do you think?
> 
> 
> 1 Unless 2 Handlers of the same name are defined in different packages and both used.

This makes sense, except that for case 1. sometimes the entire handling of the `Upon...` event must be an auxiliary function, because it is code that could be called somewhere (an example of that is applying a prepare message, which must be process after buffer in the event of a view change). I think that in this case we should stick to 1. but leave `apply...` instead of approach 2. (to differentiate functions from approach 2. from functions that are in approach 1. but that happen to be used elsewhere. Weak opinition weakly held. What is your opinion?

File: pkg/orderers/internal/common/pbftconfig.go [hunk 8]
```diff
@@ -53,5 +53,5 @@ type PBFTConfig struct {
 	ViewChangeResendPeriod time.Duration
 
 	// The current epoch number
-	epochNr tt.EpochNr
+	EpochNr tt.EpochNr
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/common/pbftconfig.go:56]: Added issue #439

File: pkg/orderers/internal/parts/goodcase/pbftgoodcase.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/parts/goodcase/pbftgoodcase.go]: See #440 for the renaming

File: pkg/orderers/internal/parts/catchup/catchup.go [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+func IncludeSegmentCheckpoint(
+	m dsl.Module,
+	state *common.State,
+	params *common.ModuleParams,
+	moduleConfig common2.ModuleConfig,
+	logger logging.Logger,
+) {
+
+	// UponResultOne with a CatchUpResponse as context,
+	// process the missing committed certificate as its hash became available.
```
[the remark this answers is not in the document]
Comment (ranchalp) [reply pkg/orderers/internal/parts/catchup/catchup.go:38]: #439

**Comment by ranchalp** [comment]
All comments converted to issues or resolved :)

**Review by matejpavlovic — APPROVED** [review]
Great job Alex!
I leave you the pleasure of clicking the "Merge" button :-)
(Just don't forget to leave a meaningful commit message of the squashed commits.)

Regarding the unified way of using handler functions, anything you said is fine with me. If a handler is also called indirectly (like applying a prepare message, or applying buffered messages) in a file where option 1. is used, I'll be happy with any of the following.
- Switching the whole file to option 2.
- Making an exception and defining just that function in option-2-style, with an option-2-name
- Defining just that function with a different name (like `apply...`)

We don't need to over-specify things, especially since there are many way more urgent issues to address. A style guide would be nice though, eventually, I'll add an issue for that.

**Pull request merged**